### ISSUE TITLE
unable to use `expect_offense` to test `on_other_file` on custom cops

### ISSUE THREAD

**Issue by karmix** [rubocop/rubocop#10437]
When the processed source is not valid Ruby syntax, [RuboCop calls `on_other_file`](https://github.com/rubocop/rubocop/blob/924b939202eddeee2516aec958a2487356f3e53c/lib/rubocop/cop/commissioner.rb#L83-L90) on each of the cops so they can investigate the file.  When unit testing cops with RSpec, the `expect_offense` helper function uses a [wrapper for `parse_source`](https://github.com/rubocop/rubocop/blob/924b939202eddeee2516aec958a2487356f3e53c/lib/rubocop/rspec/expect_offense.rb#L197-L203) which raises an exception if the processed source is not valid Ruby syntax.  The exception causes `expect_offense` to abort before starting the investigation, leaving no way to test `on_other_file` on the cop.

## Expected behavior
`expect_offense` should either never raise this exception, or should have a mechanism for telling it to suppress the exception and continue the investigation.

## Actual behavior
An exception is raised, the investigation is not started, and the cop is not called.

## Steps to reproduce the problem
Cop:
```ruby
module RuboCop
  module Cop
    module Department
      class MyCop < Base
        include RangeHelp
        MSG = 'Ye Ole Configfile is deprecated.  Convert it to DSL with ye-ole-config-converter.'
        def on_other_file
          # Never gets called w/ expect_offense.
          add_offense(source_range(processed_source.buffer, 1, 0), severity: :warning)
        end
      end
    end
  end
end
```

Spec:
```ruby
RSpec.describe RuboCop::Cop::Department::MyCop, :config do
  context 'contains old config' do
    it 'registers an offense' do
      expect_offense(<<~'SOURCE', '/my/old/Configfile')
        This is an old config.  It's content is not DSL.
        ^ Ye Ole Configfiles are deprecated.  Convert it to DSL with [...]
      SOURCE
    end
  end
end
```

## RuboCop version
[current `master`](https://github.com/rubocop/rubocop/tree/924b939202eddeee2516aec958a2487356f3e53c) branch

**Comment by karmix** [comment]
You can see [here](https://github.com/chef/cookstyle/blob/6d92d8727461c119385ca54f8ecbf5cfc7ea632f/spec/rubocop/cop/chef/deprecation/delivery_spec.rb#L31-L38) where I encountered this issue trying to test a cop in Cookstyle.  [`allow_invalid_ruby`](https://github.com/chef/cookstyle/blob/6d92d8727461c119385ca54f8ecbf5cfc7ea632f/lib/rubocop/monkey_patches/allow_invalid_ruby.rb#L15-L19) is an *ugly* monkey patch that I created to work around the issue until a better solution is available.
